Mediation take 1

katrina-wilhite opened the pull request.

katrina-wilhite committed:
Reduced mother and father race to single digits

katrina-wilhite committed:
Closes #7

katrina-wilhite committed:
Ran full simple mediation analysis for males

katrina-wilhite committed:
Ran female mediation analysis

katrina-wilhite committed:
Added code for descriptive statistics table

katrina-wilhite committed:
Changed to RMD file

katrina-wilhite committed:
Changed to RMD

katrina-wilhite committed:
Changed to RMD

katrina-wilhite committed:
Created appropriate dataframe for descriptive statistics

katrina-wilhite committed:
Finished descriptive statistics table

tarensanders commented:
As shorthand, I'm going to mark comments that are just FYIs or things to think about in the future with a :bulb:. That way you still get feedback on things like style or efficiency to implement in the future, but you don't feel like you need to go back and do it again.

tarensanders commented on simple_mediation.rmd:
````diff
@@ -0,0 +1,123 @@
+options(tinytex.verbose = TRUE)
+```
+
+```{r}
+#RQ: Does domain-specifical movement behaviour profile membership mediate the relationship between socioeconomic position and socio-emotional outcomes in children?
+##Simple mediation for males
+```
+
````
Probably no need to put this in a chunk - you want it to show as text right?

tarensanders commented on simple_mediation.rmd:
```diff
@@ -0,0 +1,123 @@
+  select(hicid, model3_trajectory_assignments, Age, Remoteness, Indigenous, SEP)
+
+# Select relevant columns from LSAC wave 6 (SEIFA varialbes), parents living at home, parent mental health, parent ethnicity)
+wave_6_relevant_data <- lsac_wave6 %>%
+  select(hicid, fcnfsad2d, fcnfsda2d, fcnfser2d, fcnfseo2d, fmoth, ffath, fsc13a1i, zf09fm, zf09ff) 
+
```
:bulb: see previous comment about `{styler}` and `{lintr}`. It seems silly, but simple things like maintaining line lengths force you to be more consistent in the way that you write code.

tarensanders commented on simple_mediation.rmd:
````diff
@@ -0,0 +1,123 @@
+```
+
+```{r}
+# Select relevant columns from male dataset (hicid, profile membership, age, remoteness, Indigenous status, SEP)
````
:bulb: From the tidyverse style guide:
> In code, use comments to explain the “why” not the “what” or “how”. 

and

> In data analysis code, use comments to record important findings and analysis decisions. If you need comments to explain what your code is doing, consider rewriting your code to be clearer.

That is, your comments don't need to repeat obvious things in code. It's obvious from the code what is being selected.

(I know this is silly, but the rules around writing concisely apply within code too)

tarensanders commented on simple_mediation.rmd:
````diff
@@ -0,0 +1,123 @@
+```{r, include=FALSE}
+options(tinytex.verbose = TRUE)
````
Just a note that it's not obvious that this is just for males. Since you called the other one `_females`, maybe add `_males` to this one? Or, consider merging into one file.

tarensanders commented on simple_mediation.rmd:
```diff
@@ -0,0 +1,123 @@
+  select(hicid, fcnfsad2d, fcnfsda2d, fcnfser2d, fcnfseo2d, fmoth, ffath, fsc13a1i, zf09fm, zf09ff) 
+
+# Update column names to make sense
+colnames(wave_6_relevant_data)[2:10] <- c("adv", "dis", "resources", "education_occupation", "mother_home", "father_home", "mental_health", "mother_race", "father_race")
```
:bulb: be a little wary of renaming this way. It relies on the columns never changing, but you can imagine a scenario where you decide you need to add a column above, but then have to remember where it was renamed.

Using tools like `rename` from {dplyr} can make this clearer, or using named vectors instead.

tarensanders commented on simple_mediation.rmd:
```diff
@@ -0,0 +1,123 @@
+# 2 mental health, 9 mother race, 125 father race
+df_males <- df_males[!(df_males$mental_health == -9), ]
+df_males <- df_males[!(df_males$mother_race == -9), ]
+df_males <- df_males[!(df_males$father_race == -9), ]
```
Just checking that you looked at the codebook (and other docs) to see if -9 had any kind of meaning. E.g., could it just be that they don't have fathers (or mothers don't know the fathers race)? If so, you could code these as a new variable rather than just removing. 125 is ~12% of your sample, so worth checking for other options before removing.

tarensanders commented on simple_mediation.rmd:
```diff
@@ -0,0 +1,123 @@
+#Reduce mother and father race to 1 digit and convert to factor 
+df_males_na_removed$mother_race <- substr(df_males_na_removed$mother_race, 1, 1) %>% 
+  as.factor()
+df_males_na_removed$father_race <- substr(df_males_na_removed$father_race, 1, 1) %>% 
```
A quick look at the tabs for this shows that some of these groups are still really tiny. E.g., group 9 has only 3 participants. I would have a look at if you can 1) collapse some of these (e.g., is the distinction between different areas of Asia important?) and 2) aggregate the smaller ones into an 'other' category. Depending on the reason you've included this, it might even be better just to go with born in aus vs not?

tarensanders commented on simple_mediation.rmd:
```diff
@@ -0,0 +1,123 @@
+  as.factor()
+
+# Covert relevant columns to factors 
+factor_columns <- c(2,4:5,7:13)
```
This makes me really nervous. There's no good way to check that these are the right columns to convert to factors, and if anything changes above it would be really easy for this not to get updated. A named vector removes this risk.

tarensanders replied:
Another option would be to avoid the loop entirely. 

```r
cols <- c("A", "B", "C")
df_males_na_removed[name] <- lapply(data[cols], factor)
```

or in dplyr:

```r
df_males_na_removed %>% mutate_at(c("A", "B", "C"), as.factor)
```

tarensanders commented on descriptive_statistics.R:
````diff
@@ -0,0 +1,163 @@
+  
+```{r include=FALSE}
+#Make Descriptive Statistics Table
+#install.packages("table1")
````
:bulb: at some point you should look into `{renv}`.

tarensanders commented on female_lasso_mediation.rmd:
````diff
@@ -0,0 +1,218 @@
+
+```{r}
+#Select exposure variables (SEP) - already standarized in LSAC
+exposure_variable <- df_females %>% 
+  select(SEP) 
+exposure_variable <- exposure_variable$SEP 
+```
````
You could replace all of this with 

```r
exposure_variable <- dplyr::pull(df_females, SEP)
```

tarensanders commented on female_lasso_mediation.rmd:
```diff
@@ -0,0 +1,218 @@
+df_outcomes <- cbind(hapsoc, hahypr, haemot, hapeer,  hacondb, hasdqtb) 
+df_outcomes <- as.data.frame(df_outcomes)
+#Make lambda grid to be used in function to find best fit model 
+lambda.grid <- seq(from = 0.4, to = 0.01, by = -0.01)
```
:bulb: see style guide for variable naming. Regardless, it's best not to mix naming conventions. Either use underscores or dots, but not both.

tarensanders commented on female_lasso_mediation.rmd:
```diff
@@ -0,0 +1,218 @@
+  x <- outcome_list$exposure
+  y <- outcome_list$outcome
+  med <- outcome_list$mediator
+  dat.filter <- assign(paste0("dat.filter.", outcome), regmed.prefilter(x, med, y, k = 10)) 
```
Why are you selecting only 10 mediators?

tarensanders commented:
Things we should discuss:

- [ ] your descriptive_statistics is a .R file, but structured as a .rmd??
- [ ] I worry about sources of truth. You have the same code in an R file and an .rmd for the mediation. 
- [ ] Code repetition across descriptive files and analysis
- [ ] Where is male lasso mediation?

katrina-wilhite committed:
Updated main script
Combined male and female into same script

katrina-wilhite committed:
Loaded male and female data

katrina-wilhite committed:
Created usable datasetes for mediation analysis

katrina-wilhite committed:
Updated RMD file so only results are shown

katrina-wilhite committed:
Moved code to "simple_mediation"

katrina-wilhite committed:
Created lasso munge

katrina-wilhite committed:
Adapated code to "mvregemd.grid" function
Combined male and female in same script

katrina-wilhite committed:
Deleted regmed code and made loop for single outcomes

katrina-wilhite committed:
Updated coding style

katrina-wilhite committed:
Summarized lasso results

katrina-wilhite committed:
Updated variable names

katrina-wilhite committed:
Added ".1" to all 0 values

katrina-wilhite committed:
Added code for single movement behaviours

katrina-wilhite committed:
Added male and female data line

katrina-wilhite committed:
Merge branch 'Mediation_Take_1' of https://github.com/katrina-wilhite/Study_3 into Mediation_Take_1

katrina-wilhite committed:
Added code for two-way interaction munge only

katrina-wilhite committed:
added code for single and two-way interactions

katrina-wilhite committed:
Updated results for male and female simple mediation

katrina-wilhite committed:
Attempted making results table

katrina-wilhite committed:
Added .1 to all 0 values

katrina-wilhite committed:
Changed lambda grid from .4 to .2

katrina-wilhite committed:
Merge branch 'Mediation_Take_1' of https://github.com/katrina-wilhite/Study_3 into Mediation_Take_1

katrina-wilhite committed:
Added ideas for table

katrina-wilhite committed:
Updated results

katrina-wilhite committed:
Fixed conflicts

Merge branch 'Mediation_Take_1' of https://github.com/katrina-wilhite/Study_3 into Mediation_Take_1

# Conflicts:
#	simple_mediation_results.rmd

katrina-wilhite committed:
Attempted to make function for results table

katrina-wilhite committed:
Made function and loop to make results table

katrina-wilhite committed:
Made simple mediation results table

katrina-wilhite committed:
Updated table name

katrina-wilhite committed:
Updated function name

katrina-wilhite committed:
Added two way interaction lasso

katrina-wilhite committed:
Cleaned up code

katrina-wilhite committed:
Organized code

katrina-wilhite committed:
Added munge code lines to script

katrina-wilhite committed:
Plotted mediation effects

katrina-wilhite committed:
Updated variable names

katrina-wilhite committed:
Created base lasso plots

katrina-wilhite committed:
Customized figures

katrina-wilhite committed:
Prepared data for model selection table

katrina-wilhite committed:
Updated table for model selection

katrina-wilhite committed:
Prepared plots for knitting

katrina-wilhite committed:
Saved results

katrina-wilhite committed:
Prepared supplementary table for knitting

katrina-wilhite committed:
Prepared table for knitting

katrina-wilhite committed:
Added puberty variables

katrina-wilhite committed:
Cleaned code so it is easier to read

katrina-wilhite committed:
Updated table to remove <.01

tarensanders commented on munging.R:
```diff
@@ -0,0 +1,49 @@
+
+munge <- function(sex) {
+  if (sex == "male"){
+    model_used = "model3_trajectory_assignments"
```
Is there a reason you've switch from `<-` to `=`?

tarensanders commented on munging.R:
```diff
@@ -0,0 +1,49 @@
+    dplyr::rename("prosocial" = "hapsoc", "hyperactivity" = "hahypr", "emotional" = "haemot", "peer" = "hapeer", "conduct" = "hacondb", "total" = "hasdqtb")
+  domsp_wave6 <- inner_join(relevant_sex_data, wave_6_relevant_data, by = "hicid")
+  df_sex <- inner_join(domsp_wave6, wave_8_relevant_data, by = "hicid")
+  df_sex[!(df_sex$mental_health == -9), ]
```
This gets printed to the console, but doesn't actually change anything?

Totally fine to have lines like this as you go interactively, but your scripts should just have the code needed to redo the analysis and not all these checks because they make it harder to follow.

tarensanders commented on munging.R:
```diff
@@ -0,0 +1,49 @@
+  df_sex <- inner_join(domsp_wave6, wave_8_relevant_data, by = "hicid")
+  df_sex[!(df_sex$mental_health == -9), ]
+  df_sex <- na.omit(df_sex)
+  df_sex[(df_sex$mother_race == 1101),]
```
See above. Also I worry that future Katrina won't know what `-9` and `1101` refer to. Comments would help.

tarensanders commented on munging.R:
```diff
@@ -0,0 +1,49 @@
+  cols <- c("prosocial", "hyperactivity", "emotional", "peer", "conduct", "total")
+  df_sex <- df_sex %>% 
+    mutate_at(c("prosocial", "hyperactivity", "emotional", "peer", "conduct", "total"), scale)
+  assign(paste0("df_",sex), envir = globalenv(), df_sex)
```
:bulb: it's not important now, but just as a learning opportunity - it's generally better to avoid casting things to the global environment. 

The reason is that if you write a function that *returns* something, then when you use that function it is clear what is going on.

E.g., at the moment to use `munge` you call `munge("male")`. That puts a variable in the global environment. But, unless you have read the code for `munge`, you do not know that and you do not know what the new variable(s) are called.

Instead, if `munge` returned a dataframe (or, a list of dataframes) then you could cast it to a new variable. The person reading your code now knows that something has been returned and what it is called.

```R
male_df <- munge("male")
```

tarensanders commented on simple_mediation.R:
```diff
@@ -1,86 +1,58 @@
+munge("female")
+
+#Run loop for all socio-emotional variables
+for (x in 16:ncol(df_male)) {
```
Be a bit careful with `16:ncol()`, because if you change something upstream which affects the number of columns this will fail silently, and is pretty hard to debug.

A list of column names can be safer.

tarensanders commented on simple_mediation.R:
```diff
@@ -1,86 +1,58 @@
+
+
+
+#I tried to make a function but I couldn't get it to work! I couldn't figure it out but it kept returning erros out on the last line. No clue why.  
```
Hard to say. Nothing is obviously wrong and I'd need the error message to be more help.

I would remove all of this though.

tarensanders commented on simple_mediation_results.rmd:
```diff
@@ -0,0 +1,85 @@
+effdecomp_list <- list(male_effdecomp_prosocial, male_effdecomp_hyperactivity, male_effdecomp_emotional, male_effdecomp_peer, male_effdecomp_conduct, male_effdecomp_total, female_effdecomp_prosocial, female_effdecomp_hyperactivity, female_effdecomp_emotional, female_effdecomp_peer, female_effdecomp_conduct, female_effdecomp_total)
+
+#Run loop for males and then females - group by males and females in kable table 
+for (i in effdecomp_list) {
```
Nice use of functions!

tarensanders commented on simple_mediation_results.rmd:
```diff
@@ -0,0 +1,85 @@
+results_table$p.2[results_table$p.2 == "0.0"] <- "<.01"
+results_table$outcome <- tools::toTitleCase(results_table$outcome)
+results_table <- results_table %>% 
+  mutate(outcome = str_replace(outcome, "Prosocial", "Prosocial Behaviour")) %>% 
```
If you use `str_replace_all` you can do all of these in one line

```R
results_table %>% 
    mutate(outcome = str_replace_all(outcome, 
    c("Prosocial" = "Prosocial Behaviour",
     "Hyperactivity" = "Hyper"
     # etc
     )))
```

tarensanders commented on lasso_munging.R:
```diff
@@ -0,0 +1,61 @@
+
+munge_lasso <- function(sex) {
+  if (sex == "male"){
+    d = df_male_domsp_weekday
```
You could just pass the dfs into the function to avoid needing this if statement

tarensanders commented on simple_mediation_females.R:
```diff
@@ -0,0 +1,2 @@
+#See "simple_mediation.R" 
```
Delete these files - no need to keep unused files

tarensanders reviewed (approved):
This looks good to me, Katrina. I don't think there's anything you really need to change - it's all just style and suggestions for the future. 

I would suggest you remove all of the redundant files, otherwise it will be hard to know where to look when you come back. Let me know when you've done that and we'll merge this in.

katrina-wilhite committed:
Added munge for general movement behaviours

katrina-wilhite committed:
Added mediation for general movement behaviours

katrina-wilhite committed:
Added results for general movement behaviours

katrina-wilhite committed:
Added munge for general movement behaviour interactions

katrina-wilhite committed:
Added mediation analysis for general movement behaviour interactions

katrina-wilhite committed:
Updated results table to include general movement behaviours

katrina-wilhite committed:
Made table clearer

katrina-wilhite committed:
Updated "0" to "0.00"

katrina-wilhite committed:
Updated table name

katrina-wilhite committed:
Add Cloudstor package

katrina-wilhite committed:
Update cloud_auth and retrieve datasets

katrina-wilhite committed:
Saved datasets to global enviornment

katrina-wilhite committed:
Removed parent occupation and education

katrina-wilhite committed:
Removed another instance of education and occupation

katrina-wilhite committed:
Removed parent occupation and education from analysis

katrina-wilhite committed:
Update columns in loop

katrina-wilhite committed:
Updated table

katrina-wilhite committed:
Removed "mother race" to allow for model to properly converge

katrina-wilhite committed:
Updated table with updated model results

katrina-wilhite committed:
Added general movement behaviors to descriptive statistics table
